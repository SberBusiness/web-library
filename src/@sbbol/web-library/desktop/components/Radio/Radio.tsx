import React, {useState, useRef, useEffect} from 'react';
import {IRadioProps} from '@sbbol/web-library/desktop/components/Radio/types';
import {classnames} from '@sbbol/web-library/common/utils/classnames/classnames';
import {EFocusSource} from '@sbbol/web-library/common/enums/EFocusSource';
import {isKey} from '@sbbol/web-library/desktop/utils/keyboard';

/** Компонент Radio. Радио-кнопка. */
export const Radio = React.forwardRef<HTMLInputElement, IRadioProps>((props, ref) => {
    const {children, className, onFocus, labelAttributes, ...inputAttributes} = props;
    const [focusSource, setFocusSource] = useState(EFocusSource.NONE);
    const labelRef = useRef<HTMLLabelElement | null>(null);
    const classNames = classnames('cssClass[radio]', {'cssClass[focusVisible]': focusSource === EFocusSource.KEYBOARD}, className);
    const labelClassNames = classnames('cssClass[label]', {'cssClass[nonempty]': !!children}, labelAttributes?.className);

    useEffect(() => {
        if (focusSource !== EFocusSource.NONE) {
            document.addEventListener('mousedown', listenMouseDown);
            document.addEventListener('keydown', listenKeyDown);
            return () => {
                document.removeEventListener('mousedown', listenMouseDown);
                document.removeEventListener('keydown', listenKeyDown);
            };
        }
    }, [focusSource]);

    const listenMouseDown = (event: MouseEvent) => {
        if (!labelRef.current?.contains(event.target as Node)) {
            setFocusSource(EFocusSource.NONE);
        }
    };

    const listenKeyDown = (event: KeyboardEvent) => {
        const key = event.code || event.keyCode;

        if (isKey(key, 'TAB')) {
            setFocusSource(EFocusSource.NONE);
        }
    };

    const handleMouseDown = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (focusSource === EFocusSource.NONE && labelRef.current !== document.activeElement) {
            setFocusSource(EFocusSource.MOUSE);
        }
        labelAttributes?.onMouseDown?.(event);
    };

    const handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource === EFocusSource.NONE) {
            setFocusSource(EFocusSource.KEYBOARD);
        }
        onFocus?.(event);
    };

    return (
        <label {...labelAttributes} className={labelClassNames} onMouseDown={handleMouseDown} ref={labelRef}>
            <input type="radio" className={classNames} onFocus={handleFocus} {...inputAttributes} ref={ref} />
            <span className="cssClass[radioIcon]" />
            <span className="cssClass[content]">{children}</span>
        </label>
    );
});

Radio.displayName = 'Radio';
