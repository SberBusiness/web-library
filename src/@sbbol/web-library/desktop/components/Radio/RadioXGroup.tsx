import React from 'react';
import {TIndentSize} from '@sbbol/web-library/desktop/common/consts/IndentConst';
import {classnames} from '@sbbol/web-library/common/utils/classnames/classnames';
import {cssClass} from '@sbbol/web-library/desktop/utils/cssClass';

/** Свойства компонента RadioXGroup. */
export interface IRadioXGroupProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Размер отступа. */
    indent?: TIndentSize;
}

/** Компонент RadioXGroup. Группа радио-кнопок с направлением по оси X. */
export const RadioXGroup: React.FC<IRadioXGroupProps> = (props) => {
    const {children, className, indent = 12, ...rest} = props;
    const classNames = classnames('cssClass[radioXGroup]', cssClass(`indent-${indent}`), className);

    return (
        <div className={classNames} role="radiogroup" {...rest}>
            {children}
        </div>
    );
};

RadioXGroup.displayName = 'RadioXGroup';
