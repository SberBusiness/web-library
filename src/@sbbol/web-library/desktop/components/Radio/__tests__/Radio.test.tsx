import React from 'react';
import renderer from 'react-test-renderer';
import {allure} from '@jest/unit/allure-report';
import {shallow} from 'enzyme';
import {Radio} from '@sbbol/web-library/desktop/components/Radio/Radio';

describe('Radio', () => {
    beforeEach(() => {
        allure.feature('Radio');
    });

    it('renders correctly', () => {
        const tree = renderer.create(<Radio>Radio</Radio>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders disabled state correctly', () => {
        const tree = renderer.create(<Radio disabled>Radio</Radio>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls onChange handler', () => {
        const handleChange = jest.fn();
        const wrapper = shallow(
            <Radio onChange={handleChange} checked>
                Radio
            </Radio>
        );

        expect(wrapper.find('input[type="radio"]').props().checked).toBe(true);
        wrapper.find('input[type="radio"]').simulate('change', {target: {checked: false}});
        expect(handleChange).toHaveBeenCalledWith({target: {checked: false}});
    });

    it('gets disabled', () => {
        const handleChange = jest.fn();
        const wrapper = shallow(
            <Radio onChange={handleChange} disabled>
                Radio
            </Radio>
        );

        expect(wrapper.find('input[type="radio"]').props().disabled).toBe(true);
    });
});
