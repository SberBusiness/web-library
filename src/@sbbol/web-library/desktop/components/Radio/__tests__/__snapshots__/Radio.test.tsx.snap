// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly 1`] = `
<label
  className="cssClass[label] cssClass[nonempty]"
  onMouseDown={[Function]}
>
  <input
    className="cssClass[radio]"
    onFocus={[Function]}
    type="radio"
  />
  <span
    className="cssClass[radioIcon]"
  />
  <span
    className="cssClass[content]"
  >
    Radio
  </span>
</label>
`;

exports[`Radio renders disabled state correctly 1`] = `
<label
  className="cssClass[label] cssClass[nonempty]"
  onMouseDown={[Function]}
>
  <input
    className="cssClass[radio]"
    disabled={true}
    onFocus={[Function]}
    type="radio"
  />
  <span
    className="cssClass[radioIcon]"
  />
  <span
    className="cssClass[content]"
  >
    Radio
  </span>
</label>
`;
